stdlib/json.gr

```diff
@@ -0,0 +1,2155 @@
+ *
+ * @param json: The JSON object to print
+ * @param format: formatting option
+ * @returns A `Result` object with either the string containing the printed JSON or an error if the input object cannot be represented in the JSON format.
+ *
+ * @example
+ * print(
+ *   Result.unwrap(
+ *     toString(
+ *       format=Custom(defaultCompactAndSafeFormat),
+ *       JsonObject([("currency", JsonString("€")), ("price", JsonNumber(99.9))])
+ *     )
+ *   )
+ * )
+ * // Output: {"currency":"\u20ac","price":99.9}"
+ * @example
+ * print(
+ *   Result.unwrap(
+ *     toString(
+ *       JsonObject([("currency", JsonString("€")), ("price", JsonNumber(99.9))]
+ *     )
+ *   )
+ * )
+ * // Output: {"currency":"€","price":99.9}
+ * @example
+ * print(
+ *   Result.unwrap(
+ *     toString(
+ *       format=Compact
+ *       JsonObject([("currency", JsonString("€")), ("price", JsonNumber(99.9))])
+ *      )
+ *   )
+ * )
+ * // Output: {"currency":"€","price":99.9}
+ * @example
+ * print(
+ *   Result.unwrap(
+ *     toString(
+ *       format=Pretty,
+ *       JsonObject([("currency", JsonString("€")), ("price", JsonNumber(99.9))])
+ *     )
+ *   )
+ * )
+ * // Output:
+ * // {
+ * // "currency": "€",
+ * // "price": 99.9
+ * //}
+ *
+ * @since v0.6.0
+ */
+provide let toString = (format=Compact, json: Json) => {
+  let buf = Buffer.make(16)
+  let writer = match (format) {
+    Pretty => makePrettyPrintJsonWriter(defaultPrettyFormat, buf),
+    Compact => makeCompactJsonWriter(buf),
+    Custom(format) => makePrettyPrintJsonWriter(format, buf),
+  }
+  let error = writer.emit(json)
+
+  match (error) {
+    None => Ok(Buffer.toString(buf)),
+    Some(e) => Err(e),
+  }
+}
+
+/**
+ * Represents errors for JSON parsing along with a human readable text message.
+ */
+provide enum JsonParseError {
+  UnexpectedEndOfInput(String),
+  UnexpectedToken(String),
+  InvalidUTF16SurrogatePair(String),
+}
+
+/**
+ * Internal data structure used during parsing.
+ */
+record JsonParserState {
+  string: String,
+  bufferParse: Buffer.Buffer,
+  mut currentCodePoint: Number,
+  mut pos: Number,
+  mut bytePos: Number,
+}
+
+let isInterTokenWhiteSpace = (codePoint: Number) => {
+  match (codePoint) {
+    0x0009 => true, // tab
+    0x000A => true, // line feed
+    0x000D => true, // carriage return
+    0x0020 => true, // space
+    _ => false,
+  }
+}
+
+let _END_OF_INPUT = -1
+
+@unsafe
+let rec readCodePoint = (bytePosition: Number, string: String) => {
+  from WasmI32 use { (+), (<) }
+
+  let strPtr = WasmI32.fromGrain(string)
+
+  let byteSize = WasmI32.load(strPtr, 4n)
+
+  let bytePositionW32 = coerceNumberToWasmI32(bytePosition)
+
+  let ptr = strPtr + 8n + bytePositionW32
+
+  if (bytePositionW32 < byteSize) {
+    let codePoint = getCodePoint(ptr)
+    tagSimpleNumber(codePoint)
+  } else {
+    _END_OF_INPUT
+  }
+}
+
+let codePointUTF8ByteCount = (usv: Number) => {
+  if (!Char.isValid(usv)) {
+    throw InvalidArgument("Invalid unicode scalar value")
```

Comment:
This should result in crashing the process?

Reply:
I think this is fine, we may want to change this to a `fail "impossible"` Thoughts? this is used internally and the program would have already errored at `getCodePoint` before hitting this.

Reply:
`fail` with impossible messages are fine, but you need to have specific ones for each fail so we can trace them down when someone inevitably hits it.